File: DotNet/Controllers/InvoiceDetailController.cs
using Microsoft.AspNetCore.Mvc;
using Vehicle_Configurator_Project.DTOS;
using Vehicle_Configurator_Project.IServices;
using Vehicle_Configurator_Project.Models;
using Vehicle_Configurator_Project.Services;

namespace Vehicle_Configurator_Project.Controllers
{
    [ApiController]
    [Route("api/invoicedetail")]
    public class InvoiceDetailController : ControllerBase
    {
        private readonly IInvoiceDetailService _invoiceDetailService;
        private readonly IAlternateComponentService _alternateComponentService;
        private readonly IHttpClientFactory _httpClientFactory;

        public InvoiceDetailController(
    IInvoiceDetailService invoiceDetailService,
    IAlternateComponentService alternateComponentService,
    IHttpClientFactory httpClientFactory)
        {
            _invoiceDetailService = invoiceDetailService;
            _alternateComponentService = alternateComponentService;
            _httpClientFactory = httpClientFactory;
        }


        [HttpGet]
        public async Task<ActionResult<List<InvoiceDetail>>> GetAllInvoiceDetails()
        {
            var list = await _invoiceDetailService.GetAllInvoiceDetailsAsync();
            return Ok(list);
        }

        [HttpPost]
        public async Task<ActionResult<InvoiceDetail>> AddInvoiceDetail([FromBody] InvoiceDetail detail)
        {
            var created = await _invoiceDetailService.CreateInvoiceDetailAsync(detail);
            return CreatedAtAction(nameof(GetInvoiceDetailsByInvoice), new { invoiceId = created.InvoiceHeader.InvId }, created);
        }

        [HttpGet("byinvoice/{invoiceId}")]
public async Task<ActionResult<List<InvoiceDetail>>> GetInvoiceDetailsByInvoice(int invoiceId)
{
    try
    {
        var list = await _invoiceDetailService.GetInvoiceDetailsByInvoiceIdAsync(invoiceId);
        if (list == null || !list.Any())
            return NotFound();

        return Ok(list);
    }
    catch (Exception ex)
    {
        // Log to console or any logger you have
        Console.Error.WriteLine($"Error fetching invoice details for invoiceId {invoiceId}: {ex}");
        // Return detailed error to client (remove in production)
        return StatusCode(500, new { error = ex.Message, stack = ex.StackTrace });
    }
}



        [HttpPost("sendemail")]
        public async Task<ActionResult<string>> SendEmailViaMicroservice([FromBody] InvoiceWrapperDTO invoiceWrapperDto)
        {
            if (invoiceWrapperDto == null || invoiceWrapperDto.InvoiceDetails == null || !invoiceWrapperDto.InvoiceDetails.Any())
                return BadRequest("Invoice details are missing or invalid.");

            // Map DTO data to the payload for Java microservice
            var headerMap = new
            {
                invId = invoiceWrapperDto.InvoiceHeader?.InvId,
                custDetails = invoiceWrapperDto.InvoiceHeader?.CustDetails,
                invDate = invoiceWrapperDto.InvoiceHeader?.InvDate,  // string already
                quantity = invoiceWrapperDto.InvoiceHeader?.Quantity,
                finalAmount = invoiceWrapperDto.InvoiceHeader?.FinalAmount,
                tax = invoiceWrapperDto.InvoiceHeader?.Tax,
                totalAmount = invoiceWrapperDto.InvoiceHeader?.TotalAmount
            };

            var modelId = invoiceWrapperDto.ModelId;

            var detailsList = new List<object>();
            foreach (var detail in invoiceWrapperDto.InvoiceDetails)
            {
                if (detail?.Component == null) continue;

                decimal price = 0;

                if (string.Equals(detail.IsAlternate, "Y", StringComparison.OrdinalIgnoreCase))
                {
                    // Assuming you have a method/service to get alternate price:
                    var alternates = await _alternateComponentService.GetByModelIdAndCompIdAsync(modelId, detail.Component.CompId ?? 0);
                    var match = alternates.FirstOrDefault();
                    if (match?.DeltaPrice != null)
                        price = match.DeltaPrice.Value;
                }
                // else price stays 0 or default

                detailsList.Add(new
                {
                    component = new
                    {
                        compId = detail.Component.CompId,
                        compName = detail.Component.CompName,
                        price = (double)price  // match Java's Double type
                    },
                    isAlternate = detail.IsAlternate
                });
            }

            var emailPayload = new
            {
                invoiceHeader = headerMap,
                invoiceDetails = detailsList
            };

            var client = _httpClientFactory.CreateClient();
            try
            {
                var response = await client.PostAsJsonAsync("http://localhost:9090/email/send", emailPayload);
                response.EnsureSuccessStatusCode();
                var respStr = await response.Content.ReadAsStringAsync();
                return Ok(respStr);
            }
            catch (Exception ex)
            {
                return StatusCode(500, $"Failed to send email: {ex.Message}");
            }
        }

    }
}

File: .NET & JAVA/JAVA/VehicleConfigurator/src/main/java/com/example/controllers/InvoiceDetailController.java
package com.example.controllers;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.example.entities.Component;
import com.example.entities.InvoiceDetail;
import com.example.entities.InvoiceWrapper;
import com.example.entities.AlternateComponent;
import com.example.repositories.AlternateComponentRepository;
import com.example.repositories.ComponentRepository;
import com.example.services.AlternateComponentService;
import com.example.services.InvoiceDetailService;
import com.example.entities.Model;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.client.RestTemplate;

@RestController
@RequestMapping("/invoice-details")
public class InvoiceDetailController {

    @Autowired
    private InvoiceDetailService invoiceDetailService;

    @Autowired
    private ComponentRepository componentRepository;

    @Autowired
    private AlternateComponentService alternateComponentService;

    @Autowired
    private AlternateComponentRepository alternateComponentRepository;

    @Autowired
    private RestTemplate restTemplate;

    @GetMapping
    public List<InvoiceDetail> getAllInvoiceDetails() {
        return invoiceDetailService.getAllInvoiceDetails();
    }

    @PostMapping
    public InvoiceDetail addInvoiceDetail(@RequestBody InvoiceDetail invoiceDetail) {
        return invoiceDetailService.createInvoiceDetail(invoiceDetail);
    }

    @GetMapping("/by-invoice/{invoiceId}")
    public List<InvoiceDetail> getInvoiceDetailsByInvoice(@PathVariable int invoiceId) {
        return invoiceDetailService.getInvoiceDetailsByInvoiceId(invoiceId);
    }

    @PostMapping("/send-email")
    public ResponseEntity<String> sendEmailViaMicroservice(@RequestBody InvoiceWrapper invoiceWrapper) {
        if (invoiceWrapper == null || invoiceWrapper.getInvoiceDetails() == null || invoiceWrapper.getInvoiceDetails().isEmpty()) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body("Invoice details are missing or invalid.");
        }

        Map<String, Object> emailPayload = new HashMap<>();

        // Prepare invoice header
        Map<String, Object> headerMap = new HashMap<>();
        if (invoiceWrapper.getInvoiceHeader() != null) {
            headerMap.put("invId", invoiceWrapper.getInvoiceHeader().getInvId());
            headerMap.put("custDetails", invoiceWrapper.getInvoiceHeader().getCustDetails());
            headerMap.put("invDate", invoiceWrapper.getInvoiceHeader().getInvDate() != null
                    ? invoiceWrapper.getInvoiceHeader().getInvDate().toString() : null);
            headerMap.put("quantity", invoiceWrapper.getInvoiceHeader().getQuantity());
            headerMap.put("finalAmount", invoiceWrapper.getInvoiceHeader().getFinalAmount());
            headerMap.put("tax", invoiceWrapper.getInvoiceHeader().getTax());
            headerMap.put("totalAmount", invoiceWrapper.getInvoiceHeader().getTotalAmount());
        }
        emailPayload.put("invoiceHeader", headerMap);

        // Prepare invoice details
        List<Map<String, Object>> detailsList = new ArrayList<>();
        int modelId = invoiceWrapper.getModelId(); // safer than accessing invoiceHeader.getModel().getModelId()

        for (InvoiceDetail detail : invoiceWrapper.getInvoiceDetails()) {
            if (detail == null || detail.getComponent() == null) continue;

            Map<String, Object> detailMap = new HashMap<>();
            Map<String, Object> compMap = new HashMap<>();

            Component component = detail.getComponent();
            compMap.put("compId", component.getCompId());
            compMap.put("compName", component.getCompName());

            if ("Y".equalsIgnoreCase(detail.getIsAlternate())) {
                List<AlternateComponent> alternates = alternateComponentService
                        .getAlternatesByComponentId(component.getCompId());

                AlternateComponent match = alternates.stream()
                	    .filter(alt -> alt.getModel() != null
                	            && alt.getModel().getModelId() == modelId)
                	    .findFirst()
                	    .orElse(null);

                if (match != null && match.getDeltaPrice() != null) {
                    compMap.put("deltaPrice", match.getDeltaPrice());
                } else {
                    compMap.put("deltaPrice", BigDecimal.ZERO);
                }
            } else {
                compMap.put("deltaPrice", BigDecimal.ZERO);
            }

            detailMap.put("component", compMap);
            detailMap.put("isAlternate", detail.getIsAlternate());
            detailsList.add(detailMap);
        }

        emailPayload.put("invoiceDetails", detailsList);

        // Send email via microservice
        String microserviceUrl = "http://localhost:9090/email/send";
        try {
            String response = restTemplate.postForObject(microserviceUrl, emailPayload, String.class);
            return ResponseEntity.ok(response);
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body("Failed to send email: " + e.getMessage());
        }
    }
}
